Simplified the Administration form

JevgenijVisockij opened the pull request:
<!-----------------------------------------------------------------------------
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows below with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/#pull-requests
------------------------------------------------------------------------------>

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Simplifies administration form
| Type?         | refacto
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Part of #16482
| How to test?  | Simplifying Advanced Parameters -> Administration form. Everything must look the same with couple exceptions: blue help box replaced with help text under input, required fields are marked as required now. Also form should throw errors now in case you input value that doesn't make sense. For example 'aaa' for Lifetime of front office cookies.

<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/21243)
<!-- Reviewable:end -->

JevgenijVisockij committed:
CS Fixer fixes

JevgenijVisockij commented:
This form has multiple fields where you can input any value and it will save successfully, but it will convert it to int above 0. 

I can change form handling in order to throw errors instead when input is not integer above 0, but this would slightly increase scope of the PR. I am not sure if it's needed.

Progi1984 commented:
@sowbiba I remove the label "Waiting for QA" as the PR is in "WIP"

JevgenijVisockij committed:
Added constraints to form

JevgenijVisockij commented:
> This pull request seems to contain new translation strings. I have summarized them below to ease up review:
> 
>     * `Admin.Notifications.Error`
>       
>       * [ ]  [`The field is invalid`](https://github.com/PrestaShop/PrestaShop/pull/21243/files#diff-9b2d1f91b8ce7fe3810efefba4206717R134)
> 
> 
> (Note: this is an automated message, but answering it will reach a real human)

I added constraints to the form so it now might throw errors, I am not sure about best error message.

JevgenijVisockij committed:
Added comment for function

matks reviewed (dismissed):
Code seems OK. Waiting for wording review

Julievrz commented on src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/Administration/GeneralType.php:
```diff
@@ -43,19 +45,52 @@ public function buildForm(FormBuilderInterface $builder, array $options)
+                    new Type(
+                        [
+                            'value' => 'numeric',
+                            'message' => $this->trans('The field is invalid', 'Admin.Notifications.Error'),
```
There is a string that already exists "The %s field is invalid." in Admin.Notifications.Error Couldn't we use this one?

Otherwise, we also have "This field is invalid." in Admin.Notifications.Error.

Progi1984 commented on src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/AdministrationController.php:
```diff
@@ -115,15 +123,43 @@ public function processUploadQuotaFormAction(Request $request)
+     *
+     * @return Response|null
+     */
+    protected function renderForm(FormInterface $generalForm, FormInterface $uploadQuotaForm, FormInterface $notificationsForm)
```
```suggestion
    protected function renderForm(FormInterface $generalForm, FormInterface $uploadQuotaForm, FormInterface $notificationsForm): ?Response
```

Progi1984 commented on src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/AdministrationController.php:
```diff
@@ -115,15 +123,43 @@ public function processUploadQuotaFormAction(Request $request)
      * @param Request $request
      *
-     * @return RedirectResponse
+     * @return Response
```
```suggestion
     * @return Response|null
```

Progi1984 commented on src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/AdministrationController.php:
```diff
@@ -96,15 +100,19 @@ public function processGeneralFormAction(Request $request)
      * @param Request $request
      *
-     * @return RedirectResponse
+     * @return Response
```
```suggestion
     * @return Response|null
```

Progi1984 commented on src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/AdministrationController.php:
```diff
@@ -77,15 +77,19 @@ public function indexAction()
      * @param Request $request
      *
-     * @return RedirectResponse
+     * @return Response
```
```suggestion
     * @return Response|null
```

Progi1984 reviewed (commented):
Some feedback

Julievrz commented on src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/Administration/GeneralType.php:
```diff
@@ -43,19 +45,52 @@ public function buildForm(FormBuilderInterface $builder, array $options)
             ])
             ->add('cookie_samesite', ChoiceType::class, [
-                'required' => true,
+                'label' => $this->trans('Cookie SameSite', 'Admin.Advparameters.Feature'),
```
This wording looks new... What does it mean? "SameSite"

Julievrz commented on src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/Administration/GeneralType.php:
```diff
@@ -43,19 +45,52 @@ public function buildForm(FormBuilderInterface $builder, array $options)
             ])
             ->add('cookie_samesite', ChoiceType::class, [
-                'required' => true,
+                'label' => $this->trans('Cookie SameSite', 'Admin.Advparameters.Feature'),
```
Are this wording and the following one new? I couldn't find them on Crowdin nor on the software. If yes, is there an issue where the wording was specified? Or can you show me a screenshot where it would appear in the software? Thanks in advance!

JevgenijVisockij replied:
I think this wording is brand new but my PR is not the one that introduces it. 

This PR introduced this setting(and wording with it I guess)
https://github.com/PrestaShop/PrestaShop/pull/20601

Here is the screenshot for where it appears.
![image](https://user-images.githubusercontent.com/47781757/99527273-34f8df00-29a5-11eb-9134-48f6083f80ed.png)

Julievrz replied:
Perfect then, thank you very much @JevgenijVisockij!

Julievrz commented:
> > This pull request seems to contain new translation strings. I have summarized them below to ease up review:
> > ```
> > * `Admin.Notifications.Error`
> >   
> >   * [ ]  [`The field is invalid`](https://github.com/PrestaShop/PrestaShop/pull/21243/files#diff-9b2d1f91b8ce7fe3810efefba4206717R134)
> > ```
> > 
> > 
> > (Note: this is an automated message, but answering it will reach a real human)
> 
> I added constraints to the form so it now might throw errors, I am not sure about best error message.

Hi! I think it is important to explain the error to the user rather than just telling him/her "The field is invalid". If I understand well the number should be superior to 0 and be an integer number. What do you think of :

"The field is invalid. Please enter a positive integer number." in Admin.Advparameters.Notifications

JevgenijVisockij committed:
Changed renderFrom to be used by index action as well

JevgenijVisockij commented:
> Some feedback

Thanks! I don't think it made sense for RenderForm to have `  * @return Response|null`  as `$twig->render()` should always return Response. So instead I modified that part and now each `process..FormAction` is correct.

JevgenijVisockij commented:
> The field is invalid. Please enter a positive integer number.

I think it makes sense, I will add that.

JevgenijVisockij committed:
Changed errors to be more descriptive

khouloudbelguith commented:
Hi @JevgenijVisockij,

You need to check the Conflicting files.

Thank you!

JevgenijVisockij committed:
Fixed the translation variables

# Conflicts:
#	src/PrestaShopBundle/Resources/views/Admin/Configure/AdvancedParameters/administration.html.twig

JevgenijVisockij commented:
> Hi @JevgenijVisockij,
> 
> You need to check the Conflicting files.
> 
> Thank you!

Done

JevgenijVisockij committed:
Updated error messages to be more accurate

Julievrz commented on src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/Administration/GeneralType.php:
```diff
@@ -43,19 +45,52 @@ public function buildForm(FormBuilderInterface $builder, array $options)
+                    new Type(
+                        [
+                            'value' => 'numeric',
+                            'message' => $this->trans('The field is invalid. Please enter an integer number greater or equal to 0.', 'Admin.Notifications.Error'),
```
```suggestion
                            'message' => $this->trans('The field is invalid. Please enter an integer greater or equal to 0.', 'Admin.Notifications.Error'),
```

Julievrz commented on src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/Administration/GeneralType.php:
```diff
@@ -43,19 +45,52 @@ public function buildForm(FormBuilderInterface $builder, array $options)
+                    new GreaterThanOrEqual(
+                        [
+                            'value' => 0,
+                            'message' => $this->trans('The field is invalid. Please enter an integer number greater or equal to 0.', 'Admin.Notifications.Error'),
```
```suggestion
                            'message' => $this->trans('The field is invalid. Please enter an integer greater or equal to 0.', 'Admin.Notifications.Error'),
```

Julievrz commented on src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/Administration/GeneralType.php:
```diff
@@ -43,19 +45,52 @@ public function buildForm(FormBuilderInterface $builder, array $options)
+                    new Type(
+                        [
+                            'value' => 'numeric',
+                            'message' => $this->trans('The field is invalid. Please enter an integer number greater or equal to 0.', 'Admin.Notifications.Error'),
```
```suggestion
                            'message' => $this->trans('The field is invalid. Please enter an integer greater or equal to 0.', 'Admin.Notifications.Error'),
```

Julievrz commented on src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/Administration/GeneralType.php:
```diff
@@ -43,19 +45,52 @@ public function buildForm(FormBuilderInterface $builder, array $options)
+                    new GreaterThanOrEqual(
+                        [
+                            'value' => 0,
+                            'message' => $this->trans('The field is invalid. Please enter an integer number greater or equal to 0.', 'Admin.Notifications.Error'),
```
```suggestion
                            'message' => $this->trans('The field is invalid. Please enter an integer greater or equal to 0.', 'Admin.Notifications.Error'),
```

Julievrz commented on src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/Administration/UploadQuotaType.php:
```diff
@@ -38,28 +40,101 @@ class UploadQuotaType extends TranslatorAwareType
+                        new Type(
+                            [
+                                'value' => 'numeric',
+                                'message' => $this->trans('The field is invalid. Please enter an integer number greater or equal to 0.', 'Admin.Notifications.Error'),
```
```suggestion
                                'message' => $this->trans('The field is invalid. Please enter an integer greater or equal to 0.', 'Admin.Notifications.Error'),
```

Julievrz commented on src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/Administration/UploadQuotaType.php:
```diff
@@ -38,28 +40,101 @@ class UploadQuotaType extends TranslatorAwareType
+                        new GreaterThanOrEqual(
+                            [
+                                'value' => 0,
+                                'message' => $this->trans('The field is invalid. Please enter an integer number greater or equal to 0.', 'Admin.Notifications.Error'),
```
```suggestion
                                'message' => $this->trans('The field is invalid. Please enter an integer greater or equal to 0.', 'Admin.Notifications.Error'),
```

Julievrz commented on src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/Administration/UploadQuotaType.php:
```diff
@@ -38,28 +40,101 @@ class UploadQuotaType extends TranslatorAwareType
+                        new Type(
+                            [
+                                'value' => 'numeric',
+                                'message' => $this->trans('The field is invalid. Please enter an integer number greater or equal to 0.', 'Admin.Notifications.Error'),
```
```suggestion
                                'message' => $this->trans('The field is invalid. Please enter an integer greater or equal to 0.', 'Admin.Notifications.Error'),
```

Julievrz commented on src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/Administration/UploadQuotaType.php:
```diff
@@ -38,28 +40,101 @@ class UploadQuotaType extends TranslatorAwareType
+                        new GreaterThanOrEqual(
+                            [
+                                'value' => 0,
+                                'message' => $this->trans('The field is invalid. Please enter an integer number greater or equal to 0.', 'Admin.Notifications.Error'),
```
```suggestion
                                'message' => $this->trans('The field is invalid. Please enter an integer greater or equal to 0.', 'Admin.Notifications.Error'),
```

Julievrz commented on src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/Administration/UploadQuotaType.php:
```diff
@@ -38,28 +40,101 @@ class UploadQuotaType extends TranslatorAwareType
+                        new Type(
+                            [
+                                'value' => 'numeric',
+                                'message' => $this->trans('The field is invalid. Please enter an integer number greater or equal to 0.', 'Admin.Notifications.Error'),
```
```suggestion
                                'message' => $this->trans('The field is invalid. Please enter an integer greater or equal to 0.', 'Admin.Notifications.Error'),
```

Julievrz commented on src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/Administration/UploadQuotaType.php:
```diff
@@ -38,28 +40,101 @@ class UploadQuotaType extends TranslatorAwareType
+                        new GreaterThanOrEqual(
+                            [
+                                'value' => 0,
+                                'message' => $this->trans('The field is invalid. Please enter an integer number greater or equal to 0.', 'Admin.Notifications.Error'),
```
```suggestion
                                'message' => $this->trans('The field is invalid. Please enter an integer greater or equal to 0.', 'Admin.Notifications.Error'),
```

Julievrz reviewed (commented):
Hi! As I told you in a previous PR, I was told to favor "integer" rather than "integer number" so I changed it here too. 
Thanks and sorry about that. Apart from that, the wording is good for me.

JevgenijVisockij committed:
Removed number from error messages

JevgenijVisockij commented:
> Hi! As I told you in a previous PR, I was told to favor "integer" rather than "integer number" so I changed it here too.
> Thanks and sorry about that. Apart from that, the wording is good for me.

I have added the changes, thanks!

Julievrz commented on src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/Administration/NotificationsType.php:
```diff
@@ -40,13 +40,16 @@ public function buildForm(FormBuilderInterface $builder, array $options)
             ->add('show_notifs_new_orders', SwitchType::class, [
-                'required' => true,
+                'label' => $this->trans('Show notifications for new orders', 'Admin.Advparameters.Feature'),
+                'help' => $this->trans('This will display notifications when new orders are made in your shop.', 'Admin.Advparameters.Help'),
```
@JevgenijVisockij @LouiseBonnard Don't you think that we could get rid of the following help texts? After reading the PR again, I feel like the label speaks for itself.

"This will display notifications when new orders are made in your shop."
"This will display notifications every time a new customer registers in your shop."
"This will display notifications when new messages are posted in your shop."

LouiseBonnard replied:
Nice catch, @Julievrz, I'm all for it! We could even go further by making the information notification looks like in #22193 and, perhaps, by shortening the label names to sound less redundant, what do you think?

Julievrz replied:
I agree with your idea of making the information notification looks like in #22193. Regarding your second idea, maybe we could have checkboxes that would allow the merchant to choose the notifications he/she wants to receive. Something like:
![image](https://user-images.githubusercontent.com/70583503/103783939-a239ff80-5039-11eb-8d25-53f047654e2a.png) 

If it is too long or complicated to implement, we could keep the current "Yes/No" buttons, by shortening the labels and putting "Show notifications for..." at the top. 

Not sure if I am being clear, let me know if you need me to specify something. 😅 

@prestascott @TristanLDD Wdyt?

Julievrz replied:
@LouiseBonnard I think this will be addressed in the issue I discussed this morning with Tristan, by implementing a toggle but this will be addressed in another issue as I think it is out of the scope of this PR.

LouiseBonnard replied:
Sure, feel free to open a new issue to address this part if you prefer! We'll just have to prioritize it along with this one to make sure the user experience is fully improved on the page. ;-)

JevgenijVisockij commented:
WIP before this https://github.com/PrestaShop/PrestaShop/pull/21553/files#r554538559 is resolved.

Julievrz commented:
I validate the wording for this PR as my last suggestion is out of scope and will be addressed in another issue. Thank you @JevgenijVisockij

JevgenijVisockij committed:
Restoring old administration controller

JevgenijVisockij committed:
Adding error messages for upload quota

Julievrz commented on src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/AdministrationController.php:
```diff
@@ -184,4 +187,71 @@ protected function getNotificationsFormHandler(): FormHandlerInterface
+        switch ($error->getErrorCode()) {
+            case UploadQuotaDataProvider::ERROR_NOT_NUMERIC_OR_LOWER_THEN_0:
+                return $this->trans(
+                    '"%s is invalid. Please enter an integer greater or equal to 0."',
```
```suggestion
                    '%s is invalid. Please enter an integer greater or equal to 0.',
```

Julievrz replied:
I'm not sure if you added the "" on purpose?

JevgenijVisockij replied:
Yeah I think it was incorrect, thanks!

Progi1984 commented on src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/Administration/UploadQuotaDataProvider.php:
```diff
@@ -0,0 +1,101 @@
+     *
+     * @return array Array of errors if any
+     */
+    private function validate(array $data)
```
```suggestion
    private function validate(array $data): array
```

Progi1984 commented on src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/Administration/UploadQuotaType.php:
```diff
@@ -30,36 +30,116 @@
+    public const FIELD_MAX_SIZE_DOWNLOADABLE_FILE = 'max_size_downloadable_product';
+    public const FIELD_MAX_SIZE_PRODUCT_IMAGE = 'max_size_product_image';
+
+
```
```suggestion
```

Progi1984 reviewed (commented):
Some feedbacks (check PHPStan ;))

JevgenijVisockij committed:
Update src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/Administration/UploadQuotaType.php

Removed uneeded line

Co-authored-by: Progi1984 <progi1984@gmail.com>

JevgenijVisockij committed:
Changed the way errors passed to controller

Progi1984 commented on src/PrestaShopBundle/Form/Exception/DataProviderError.php:
```diff
@@ -0,0 +1,85 @@
+    }
+
+    /**
+     * @param int|null $languageId
```
```suggestion
     * @return int|null
```

Progi1984 commented on src/PrestaShopBundle/Form/Exception/DataProviderException.php:
```diff
@@ -0,0 +1,54 @@
+     */
+    private $dataProviderErrors;
+
+    public function __construct($message = "", $code = 0, Throwable $previous = null, ?DataProviderErrorCollection $dataProviderErrors = null)
```
```suggestion
    public function __construct($message = '', $code = 0, Throwable $previous = null, ?DataProviderErrorCollection $dataProviderErrors = null)
```

Progi1984 reviewed (commented):
Some feedbacks

JevgenijVisockij committed:
Update src/PrestaShopBundle/Form/Exception/DataProviderException.php

Changed message to use '' instead of ""

Co-authored-by: Progi1984 <progi1984@gmail.com>

JevgenijVisockij committed:
Update src/PrestaShopBundle/Form/Exception/DataProviderError.php

Fixed return value

Co-authored-by: Progi1984 <progi1984@gmail.com>

JevgenijVisockij commented:
> Some feedbacks

Added them, thanks!

JevgenijVisockij committed:
Added error messages for general types

JevgenijVisockij commented on src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/Administration/GeneralDataProvider.php:
```diff
@@ -0,0 +1,95 @@
+            $errors->add(new DataProviderError(FormDataProvider::ERROR_NOT_NUMERIC_OR_LOWER_THEN_0, GeneralType::FIELD_BACK_COOKIE_LIFETIME));
+        }
+
+        if (!$errors->isEmpty()) {
```
Before data provider would always return errors, here I am throwing exception with those errors inside the exception. 

On one hand, I think it makes way more sense to throw exception and catch is when something goes wrong, rather then returning array of errors.

On the other this requires exception with additional param, I am not sure if that's a bad thing. Also this means that that I am changing the behaviour that exists in other options form. As part of form simplification I could change other option forms to behave this way as well, but this is a big change so I need other oppinions before I proceed with it. 

Also if we stick with returning array of errors, should I return array of DTO's or a Collection? While I like collection more, using collection breaks general behaviour that might at this point be expected of options forms.

matks replied:
I feel like
```
throw new DataProviderException('Administration general data is invalid', 0, null, $errors);
```
is actually quite a nice compromise 👍  between the "throw exception" rule and the "gather all errors" behavior

I like it like this, but I'd be happy to hear more opinions. @jolelievre @zuk3975 ?

zuk3975 replied:
If I got it right, the problem actually comes from allowing to set these invalid values in configuration in first place, but we are trying to validate something in a form DataProvider (a "read" operation) after its already wrong in db?
So if you ask me it seems all conceptually wrong (but I suppose we don't have any better way to handle this?)

Anyway, for mee too - throwing exception with a list of errors seems better than passing them around manually :+1:

zuk3975 commented on src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/Administration/FormDataProvider.php:
```diff
@@ -35,6 +35,8 @@
  */
 final class FormDataProvider implements FormDataProviderInterface
 {
+    public const ERROR_NOT_NUMERIC_OR_LOWER_THEN_0 = 1;
```
```suggestion
    public const ERROR_NOT_NUMERIC_OR_LOWER_THAN_ZERO = 1;
```

JevgenijVisockij replied:
Good point, thanks!

JevgenijVisockij committed:
Replaced constant to avoid using number

JevgenijVisockij committed:
CS fixer fixes

khouloudbelguith commented:
Hi @JevgenijVisockij,

I found some issues:

**First issue**
1. In the BO > Advanced Parameters > Administration page
2. Change the `Lifetime of back-office cookies` to `9223372036854775807`
3. Save
4. See error
![image](https://user-images.githubusercontent.com/16067358/107392111-a3a78d80-6af9-11eb-995f-6ad2db8d0dcb.png)

All the BO is NOK.

**Second issue**
1. Install the shop with `HTTP`
2. In the BO > Advanced Parameters > Administration page
3. Set the ` Cookie SameSite` to `None`
4. Save 
5. See the error, a successful alert is displayed without changing the value

**Without PR**
We have an alert 
![image](https://user-images.githubusercontent.com/16067358/107392798-55df5500-6afa-11eb-8f0d-7c114f097ce7.png)

Thank you!

JevgenijVisockij committed:
Added max number for cookie lifetime

JevgenijVisockij committed:
Fixed cookie samesite not throwing error

JevgenijVisockij commented:
>     3\. Cookie

Hey! Thanks for Q&A, fixed both issues.

matks commented on src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/Administration/GeneralDataProvider.php:
```diff
@@ -0,0 +1,147 @@
+     * If you set cookie lifetime value too high there can be multiple problems.
+     * Hours are converted to seconds, so int might be turned to float if it's way to high.
+     * Cookie classes crashes if lifetime goes beyond year 9999, there are probably multiple other things.
+     * So we need to set some sort of max value. 100 years seems like a lifetime beyond reasonable use.
```
> 100 years seems like a lifetime beyond reasonable use

yes 📘 😄  if a shop powered by PrestaShop is still running in 100 years, we can say this code was a success !

matks commented on src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/AdministrationController.php:
```diff
@@ -149,13 +158,15 @@ protected function processForm(Request $request, FormHandlerInterface $formHandl
+            try {
+                $formHandler->save($data);
+            } catch (DataProviderException $e) {
+                $this->flashErrors($this->getErrorMessages($e->getDataProviderErrors()));
```
In other controllers we have used a fallback error message when none is found
```
$this->getErrorMessageForException($e, $this->getErrorMessages($e))
```

https://github.com/PrestaShop/PrestaShop/blob/develop/src/PrestaShopBundle/Controller/Admin/Sell/Address/AddressController.php

Wdyt ? I see you rather throw a ErrorMessageNotFound instead

JevgenijVisockij replied:
Now that you mention it, I guess it would make sense to add default error.

It allows you to throw exception anywhere during the process with some custom error and process will still work.

matks commented on src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/AdministrationController.php:
```diff
@@ -184,4 +195,105 @@ protected function getNotificationsFormHandler(): FormHandlerInterface
+         * Maybe move labels to some other service and then retrieve them in both UploadQuotaType and here.
+         */
+        switch ($fieldName) {
+            case UploadQuotaType::FIELD_MAX_SIZE_ATTACHED_FILES:
```
What this little "map of error messages" inside UploadQuotaType then create a `public static function getErrorMessageMap()` function in UploadQuotaType ?

JevgenijVisockij replied:
Great idea, but static won't work because I am doing translation here and you can't reuse this->trans in static context :/

matks commented on src/PrestaShopBundle/Controller/Exception/FieldNotFoundException.php:
```diff
@@ -0,0 +1,40 @@
+ *
+ * Thrown when field for error message is not found
+ */
+class FieldNotFoundException extends DomainException
```
I think Domain Exceptions are meant to be thrown from within the domain, in src/Core/Domain

Here we are in the application layer 🤔

JevgenijVisockij replied:
Oh should have understood from naming :D Fixed

matks commented on src/PrestaShopBundle/Form/Exception/DataProviderError.php:
```diff
@@ -0,0 +1,85 @@
+
+namespace PrestaShopBundle\Form\Exception;
+
+class DataProviderError
```
Maybe we can improve the naming?

Here it's called `DataProviderError` because it's "an error returned by the DataProvider". But tomorrow we could decide to return this from another class / kind of class so the naming would become wrong.

Here are some suggestions
- ConfigurationFormValidationError => because it's errors found while validating the data input inside a configuration form
- InvalidConfigurationDataError
- WrongConfigurationDataError
- ConfigurationDataValidationError

JevgenijVisockij replied:
Thanks for multiple suggestions! Yeah what you are saying makes sense.

I like InvalidConfigurationDataError, it says more what went wrong than where it went wrong thus making it more universal. So implemented that.

matks commented on src/PrestaShopBundle/Resources/views/Admin/Configure/AdvancedParameters/administration.html.twig:
```diff
@@ -93,28 +60,7 @@
-                </div>
-              </div>
-              {{ form_rest(uploadQuotaForm) }}
+              {{ form_widget(uploadQuotaForm) }}
```
This is the part I like in this PR ❤️

JevgenijVisockij replied:
Yeah removing code is one of the more satisfying parts of programming :D

zuk3975 commented on src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/Administration/FormDataProvider.php:
```diff
@@ -35,6 +35,10 @@
  */
 final class FormDataProvider implements FormDataProviderInterface
 {
+    public const ERROR_NOT_NUMERIC_OR_LOWER_THEN_ZERO = 1;
```
THEN -> **THAN**

JevgenijVisockij replied:
Good catch, thanks!

Julievrz commented on src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/AdministrationController.php:
```diff
@@ -184,4 +195,105 @@ protected function getNotificationsFormHandler(): FormHandlerInterface
+                );
+            case FormDataProvider::ERROR_COOKIE_LIFETIME_MAX_NUMBER_EXCEEDED:
+                return $this->trans(
+                    '%s is invalid. Please enter number lower then %s.',
```
```suggestion
                    '%s is invalid. Please enter a number lower than %s.',
```

Does this number have to be an integer? If yes, maybe we should replace "number" with "integer" for more consistency. Wdyt?

JevgenijVisockij replied:
Yes it's an integer, i changeed wording to
  `` %s is invalid. Please enter an integer lower than %s.``
  Thanks for suggestion!

PierreRambaud commented on src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/Administration/GeneralDataProvider.php:
```diff
@@ -0,0 +1,147 @@
+    private function validate(array $data): void
+    {
+        $errors = new DataProviderErrorCollection();
+        $frontOfficeLifeTimeCookie = $data[GeneralType::FIELD_FRONT_COOKIE_LIFETIME];
+        $backOfficeLifeTimeCookie = $data[GeneralType::FIELD_BACK_COOKIE_LIFETIME];
```
```suggestion
        $frontOfficeLifeTimeCookie = $data[GeneralType::FIELD_FRONT_COOKIE_LIFETIME];
        $backOfficeLifeTimeCookie = $data[GeneralType::FIELD_BACK_COOKIE_LIFETIME];
```

You assume data exists all the time? :thinking: 
A validator have to test if data are `isset`

JevgenijVisockij replied:
Good point. Thanks! Do you think it would be useful to throw error in that case? I don't see it happening elsewhere in PS and I don't see value of that in form. But if data provider to be used somewhere else it could be useful.

PierreRambaud replied:
Just an `isset` call should be enough :)

JevgenijVisockij replied:
Done

PierreRambaud commented on src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/Administration/GeneralDataProvider.php:
```diff
@@ -0,0 +1,147 @@
+        }
+
+        if ($frontOfficeLifeTimeCookie > self::MAX_COOKIE_VALUE) {
+            $errors->add(new DataProviderError(FormDataProvider::ERROR_NOT_NUMERIC_OR_LOWER_THEN_ZERO, GeneralType::FIELD_FRONT_COOKIE_LIFETIME));
```
```suggestion
            $errors->add(new DataProviderError(FormDataProvider::ERROR_MAX_COOKIE_VALUE, GeneralType::FIELD_FRONT_COOKIE_LIFETIME));
```

PierreRambaud commented on src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/Administration/GeneralDataProvider.php:
```diff
@@ -0,0 +1,147 @@
+        }
+
+        if ($backOfficeLifeTimeCookie > self::MAX_COOKIE_VALUE) {
+            $errors->add(new DataProviderError(FormDataProvider::ERROR_NOT_NUMERIC_OR_LOWER_THEN_ZERO, GeneralType::FIELD_BACK_COOKIE_LIFETIME));
```
```suggestion
            $errors->add(new DataProviderError(FormDataProvider::ERROR_MAX_COOKIE_VALUE, GeneralType::FIELD_BACK_COOKIE_LIFETIME));
```

PierreRambaud commented on src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/Administration/UploadQuotaDataProvider.php:
```diff
@@ -0,0 +1,100 @@
+    private function validate(array $data): void
+    {
+        $errors = new DataProviderErrorCollection();
+        $maxSizeAttachedFile = $data[UploadQuotaType::FIELD_MAX_SIZE_ATTACHED_FILES];
```
Same as previous, you have to test if data exist in array

PierreRambaud reviewed (commented):
Few comments :)

JevgenijVisockij committed:
Changed wording to reflect that it's integer not number

JevgenijVisockij committed:
Changed "a" to "an" for integer

JevgenijVisockij committed:
checking for isset and some changes to errors

JevgenijVisockij committed:
A typo

JevgenijVisockij committed:
Missed changing from then to than in one place, so adding it.

JevgenijVisockij committed:
Changed parent for field not found exception

JevgenijVisockij committed:
Renamed DataProviderError to InvalidConfigurationDataError

JevgenijVisockij committed:
Replaced ErrorMessageNotFoundException with default error

Julievrz commented on src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/AdministrationController.php:
```diff
@@ -184,4 +195,109 @@ protected function getNotificationsFormHandler(): FormHandlerInterface
+                );
+            case FormDataProvider::ERROR_COOKIE_SAMESITE_NONE:
+                return $this->trans(
+                    'The SameSite=None is only available in Secure mode.',
```
```suggestion
                    'The SameSite=None is only available in secure mode.',
```

Julievrz commented on src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/Administration/NotificationsType.php:
```diff
@@ -40,13 +40,16 @@ public function buildForm(FormBuilderInterface $builder, array $options)
             ->add('show_notifs_new_customers', SwitchType::class, [
-                'required' => true,
+                'label' => $this->trans('Show notifications for new customers', 'Admin.Advparameters.Feature'),
+                'help' => $this->trans('This will display notifications every time a new customer registers in your shop.', 'Admin.Advparameters.Help'),
```
I think we should get rid of this help text

Julievrz commented on src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/Administration/NotificationsType.php:
```diff
@@ -40,13 +40,16 @@ public function buildForm(FormBuilderInterface $builder, array $options)
             ->add('show_notifs_new_messages', SwitchType::class, [
-                'required' => true,
+                'label' => $this->trans('Show notifications for new messages', 'Admin.Advparameters.Feature'),
+                'help' => $this->trans('This will display notifications when new messages are posted in your shop.', 'Admin.Advparameters.Help'),
```
Same here, I don't think this help message is worth keeping.

Julievrz commented on src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/Administration/NotificationsType.php:
```diff
@@ -40,13 +40,16 @@ public function buildForm(FormBuilderInterface $builder, array $options)
             ->add('show_notifs_new_orders', SwitchType::class, [
-                'required' => true,
+                'label' => $this->trans('Show notifications for new orders', 'Admin.Advparameters.Feature'),
+                'help' => $this->trans('This will display notifications when new orders are made in your shop.', 'Admin.Advparameters.Help'),
```
Here we can remove the help text, I think.

JevgenijVisockij replied:
Removed all 3 help texts

Julievrz replied:
Thank you!

Julievrz reviewed (commented):
A few comments

JevgenijVisockij committed:
Update src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/AdministrationController.php

Made secure non capitalised

Co-authored-by: Julie Varisellaz <70583503+Julievrz@users.noreply.github.com>

JevgenijVisockij committed:
Removed no longer needed help texts

JevgenijVisockij committed:
Added checking for isset before throwing errors

JevgenijVisockij committed:
CS fixer fixes

JevgenijVisockij committed:
Removed uneeded tag about thrown exception

JevgenijVisockij committed:
Simplified the Administration form

# Conflicts:
#	src/PrestaShopBundle/Resources/views/Admin/Configure/AdvancedParameters/administration.html.twig

# Conflicts:
#	src/PrestaShopBundle/Resources/views/Admin/TwigTemplateForm/prestashop_ui_kit.html.twig

Progi1984 commented on src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/AdministrationController.php:
```diff
@@ -184,4 +194,109 @@ protected function getNotificationsFormHandler(): FormHandlerInterface
+     * @param InvalidConfigurationDataError $error
+     *
+     * @return string
+     *
+     * @throws FieldNotFoundException
+     */
```
```suggestion
     */
```

khouloudbelguith commented:
Hi @JevgenijVisockij,

It is ok :heavy_check_mark: 
https://drive.google.com/file/d/1lcfs7r4Ch_2E7PhLKHzCSYD0P6ISfJtz/view?usp=sharing

Thank you!

Progi1984 commented:
Thanks @JevgenijVisockij & @khouloudbelguith
